src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,193 @@
+                  ValueID{attribute_vector->get(attribute_vector_index)} >= upper_bound) {
+                matching_pos_list->push_back(RowID{chunk_id, attribute_vector_index});
+              }
+              break;
+            case ScanType::OpLessThan:
+              if (ValueID{attribute_vector->get(attribute_vector_index)} < lower_bound) {
+                matching_pos_list->push_back(RowID{chunk_id, attribute_vector_index});
+              }
+              break;
+            case ScanType::OpLessThanEquals:
+              if (ValueID{attribute_vector->get(attribute_vector_index)} < upper_bound) {
+                matching_pos_list->push_back(RowID{chunk_id, attribute_vector_index});
+              }
+              break;
+            case ScanType::OpGreaterThan:
+              if (ValueID{attribute_vector->get(attribute_vector_index)} >= upper_bound) {
+                matching_pos_list->push_back(RowID{chunk_id, attribute_vector_index});
+              }
+              break;
+            case ScanType::OpGreaterThanEquals:
+              if (ValueID{attribute_vector->get(attribute_vector_index)} >= lower_bound) {
+                matching_pos_list->push_back(RowID{chunk_id, attribute_vector_index});
+              }
+              break;
+          }
+        }
+
+        return;
+      }
+
+      auto reference_segment = std::dynamic_pointer_cast<ReferenceSegment>(segment);
+      if (reference_segment != nullptr) {
+        const auto& referenced_pos_list = reference_segment->pos_list();
+        const auto& table_referenced_by_segment = reference_segment->referenced_table();
+        for (const RowID& referenced_pos : *referenced_pos_list) {
+          const auto& referenced_segment =
+              table_referenced_by_segment->get_chunk(referenced_pos.chunk_id).get_segment(_column_id);
+          // We are using the [] operator here because we want to avoid checking for segment type again.
+          const auto& value = type_cast<Type>((*referenced_segment)[referenced_pos.chunk_offset]);
+          if (matches_scan_criterion(value, type_cast<Type>(_search_value), _scan_type)) {
+            // Avoid ReferenceSegments to point to other ReferenceSegments.
+            if (!is_referenced_table_replaced) {
+              referenced_table = table_referenced_by_segment;
+              is_referenced_table_replaced = true;
+            }
+            matching_pos_list->push_back(referenced_pos);
+          }
+        }
+      }
+      return;
+    });
+  }
+
+  auto table = std::make_shared<Table>();
+
+  // Create table column names and types according to input table.
+  for (ColumnID column_id{0}; column_id < _input_table_left()->column_count(); ++column_id) {
+    table->add_column_definition(_input_table_left()->column_name(column_id),
+                                 _input_table_left()->column_type(column_id));
+  }
+
+  auto& chunk = table->get_chunk(ChunkID{0});
```

Comment:
creating exactly one chunk will cause problems for results with more than 2^32 elements because ChunkOffsets are 32 bits, if that ever happens